onnxruntime/contrib_ops/rocm/bert/tunable_op.h

```diff
@@ -77,31 +123,43 @@ class TunableOp {
     tuning_ = false;
   }
 
-  virtual ~TunableOp() {}
+  virtual ~TunableOp() = default;
 
  protected:
-  std::vector<std::unique_ptr<Op>> ops_;
+  std::vector<Op<ParamsT>> ops_;
 
  private:
-  virtual bool Condition(const OpParams* op_params) = 0;
-
-  int FindFastest(const OpParams* op_params) {
-    assert(ops_.size() > 0);
-    float min_time = ops_[0]->Profile(op_params);
-    int id = 0;
-    for (int i = 1; i < ops_.size(); i++) {
-      float time = ops_[i]->Profile(op_params);
+  // Whether we should tune for this input
+  virtual bool Condition(const ParamsT* /*params*/) {
```

Comment:
This default behavior may not be applicable to the derived class and may accidentally be used. Shall it keep it as a pure virtual function?

Reply:
IMO, for a `TunableOp`, we have at least a generic impl (say, rocblas, miopen or current existing sequence of kernel launches) and multiple specialized impls (say, fused). Then the tuning process will generally be supported. Otherwise, we will be leaking the checking to the call site, and it indicate a bigger problem of our `TunableOp`.

So, for our `TunableOp`, we shall always wrap the code we are targeting to replace, and keep it as the fallback, so our caller can always call it blindly and it will not break with random input.